# mavis

MAVIS workflow, annotation of structural variants. An application framework for the rapid generation of structural variant consensus, able to visualize the genetic impact and context as well as process both genome and transcriptome data. The workflow runs each MAVIS action as a WDL task; this replaces the MAVIS default method, of submitting actions directly to a computing cluster.

## Overview

## Dependencies

* [mavis 2.2.6](http://mavis.bcgsc.ca/)


## Usage

### Cromwell
```
java -jar cromwell.jar run mavis.wdl --inputs inputs.json
```

### Inputs

#### Required workflow parameters:
Parameter|Value|Description
---|---|---
`donor`|String|Donor id
`inputBAMs`|Array[BamData]|Collection of alignment files with indexes and metadata
`svData`|Array[SvData]|Collection of SV calls with metadata
`generateConfigScript.arribaConverter`|String|Path to arriba conversion script
`generateConfigScript.modules`|String|Environment modules for the task
`config.referenceGenome`|String|Path to reference FASTA file
`config.annotations`|String|Path to annotations JSON file for MAVIS
`config.masking`|String|Path to reference masking file in .tab format
`config.dgvAnnotations`|String|Path to reference Database of Genome Variants (DGV) annotations
`config.alignerReference`|String|Path to reference in 2bit (compressed) format, for the Mavis aligner
`config.templateMetadata`|String|Chromosome Band Information, used for visualization
`config.modules`|String|Environment modules for the task
`setup.modules`|String|Environment modules for the task
`validate.modules`|String|Environment modules for the task
`annotate.modules`|String|Environment modules for the task
`pairing.modules`|String|Environment modules for the task
`summary.modules`|String|Environment modules for the task
`zipResults.modules`|String|Environment modules for the task


#### Optional workflow parameters:
Parameter|Value|Default|Description
---|---|---|---
`libTypeMap`|Map[String,String]|{"WT": "transcriptome", "MR": "transcriptome", "WG": "genome"}|Mapping from library type strings to Mavis library types


#### Optional task parameters:
Parameter|Value|Default|Description
---|---|---|---
`generateConfigScript.jobMemory`|Int|12|Memory for the task, in gigabytes
`generateConfigScript.timeout`|Int|24|Timeout for the task, in hours
`config.jobMemory`|Int|12|Memory for the task, in gigabytes
`config.timeout`|Int|24|Timeout for the task, in hours
`setup.mavisAligner`|String|"blat"|String identifying the aligner
`setup.mavisDrawFusionOnly`|String|"False"|Mavis parameter: Must be 'True' or 'False'
`setup.minClusterPerFile`|Int|5|Minimum number of clusters per file
`setup.drawNonSynonymousCdnaOnly`|String|"False"|Mavis parameter: Must be 'True' or 'False'
`setup.mavisUninformativeFilter`|String|"True"|Mavis parameter: Must be 'True' or 'False'
`setup.mavisMaxFiles`|Int|200|Maximum number of chunks for parallelization
`setup.jobMemory`|Int|32|Memory for the task, in gigabytes
`setup.timeout`|Int|24|Timeout for the task, in hours
`validate.jobMemory`|Int|32|Memory for the task, in gigabytes
`validate.timeout`|Int|24|Timeout for the task, in hours
`annotate.jobMemory`|Int|32|Memory for the task, in gigabytes
`annotate.timeout`|Int|24|Timeout for the task, in hours
`pairing.jobMemory`|Int|32|Memory for the task, in gigabytes
`pairing.timeout`|Int|24|Timeout for the task, in hours
`summary.jobMemory`|Int|32|Memory for the task, in gigabytes
`summary.timeout`|Int|24|Timeout for the task, in hours
`zipResults.jobMemory`|Int|12|Memory for the task, in gigabytes
`zipResults.timeout`|Int|24|Timeout for the task, in hours


### Outputs

Output | Type | Description
---|---|---
`results`|File|ZIP archive file


## Commands
 This section lists command(s) run by mavis workflow
 
 * Running mavis
 
 MAVIS annotates structural variants for WG and WT experiments
 
 Configuring is done using custom python code which produces a configuration script ran at the next step
 
 ```
    See generateConfigScript task in mavis.wdl 
 
 ```
 
 Run Configure scripts:
 
 ```
     export MAVIS_REFERENCE_GENOME=REF_GENOME
     export MAVIS_ANNOTATIONS=ANNOTATIONS
     export MAVIS_MASKING=MASKING
     export MAVIS_DGV_ANNOTATION=DVG_ANNOTATIONS
     export MAVIS_ALIGNER_REFERENCE=ALIGN_REFERENCE
     export MAVIS_TEMPLATE_METADATA=TEMPLATE_METADATA
     chmod +x CONFIG_SCRIPT
     CONFIG_SCRIPT
 ```
 
 Setup mavis:
 
 ```
     set -euo pipefail
     unset LD_LIBRARY_PATH
     unset LD_LIBRARY_PATH_modshare
     export MAVIS_ALIGNER='MAVIS_ALIGNER'
     export MAVIS_DRAW_FUSIONS_ONLY=MAVIS_DRAW_FUSION_ONLY
     export MAVIS_MAX_FILES=MAVIS_MAX_FILE
     export DRAW_NON_SYNONYMOUS_CDNA_ONLY=DRAW_NONSYNOMOUS_CDNA_ONLY
     export min_clusters_per_file=MIN_CLUSTER_PER_FILE
     export MAVIS_UNINFORMATIVE_FILTER=MAVIS_UNINFORMATIVE_FILTER
     mavis setup CONFIG_FILE -o .
     # Try to get batch ID from the "build.cfg" file generated by setup
     grep MS_batch build.cfg | grep -v \] | sed s/.*-// | tail -n 1 | sed s/\n// > BATCH_NAME || 
       echo "unknown_batch" > BATCH_NAME
 ```
 
 Post-process and launch bash scripts (used to modify multiple scripts)
 
 ```
     set -euo pipefail
     sed -i "s|^[[:space:]]*cd .*||g" SCRIPT # do not change the working directory
     sed -i "s|^[[:space:]]*echo .*||g" SCRIPT # do not echo start/finish times
     sed -i "s|--inputs .*|--inputs IN_FILES \\\\|g" SCRIPT
     sed -i "s|--output .*|--output .|g" SCRIPT
     chmod +x SCRIPT
 
     SCRIPT
 ```
 
 Package the results into a .zip archive
 
 ```
     set -euo pipefail
     mkdir OUT_PREFIX
     cp -t OUT_PREFIX INPUT_FILES
     zip -qr OUT_PREFIX.zip OUT_PREFIX
 
 ```
 
 ## Support

For support, please file an issue on the [Github project](https://github.com/oicr-gsi) or send an email to gsi@oicr.on.ca .

_Generated with generate-markdown-readme (https://github.com/oicr-gsi/gsi-wdl-tools/)_
